tools/rtpsrelay/RelayHandler.cpp

```diff
@@ -431,6 +366,132 @@ bool VerticalHandler::parse_message(OpenDDS::RTPS::MessageParser& message_parser
   return true;
 }
 
+HorizontalHandler::HorizontalHandler(ACE_Reactor* reactor,
+                                     const AssociationTable& association_table)
+  : RelayHandler(reactor, association_table)
+  , vertical_handler_(nullptr)
+{}
+
+void HorizontalHandler::enqueue_message(const std::string& addr,
+                                        const GuidSet& guids,
+                                        ACE_Message_Block* msg)
+{
+  // Determine how many guids we can pack into a UDP message.
+  const auto max_guids = (65536 - msg->length()) / sizeof (OpenDDS::DCPS::RepoId) - 1;
+
+  auto pos = guids.begin();
+  const auto limit = guids.end();
+
+  while (pos != limit) {
+    ACE_Message_Block* header_block = new ACE_Message_Block(256);
+    OpenDDS::DCPS::Serializer ser(header_block);
+
+    RelayHeader relay_header;
+    auto& to = relay_header.to();
+    while (pos != limit && to.size() < max_guids) {
+      to.push_back(guid_to_guid(*pos));
+      ++pos;
+    }
+    ser << relay_header;
+
+    header_block->cont(msg);
+    RelayHandler::enqueue_message(addr, header_block);
+  }
+}
+
+void HorizontalHandler::process_message(const ACE_INET_Addr&,
+                                        const OpenDDS::DCPS::MonotonicTimePoint&,
+                                        ACE_Message_Block* msg)
+{
+  OpenDDS::RTPS::MessageParser mp(*msg);
+
+  const size_t size_before_header = mp.remaining();
+
+  RelayHeader relay_header;
+  if (!(mp >> relay_header)) {
+    ACE_ERROR((LM_ERROR, "(%P|%t) %N:%l ERROR: RelayHandler::handle_input failed to deserialize Relay header\n"));
+    return;
+  }
+
+  const size_t size_after_header = mp.remaining();
+
+  msg->rd_ptr(size_before_header - size_after_header);
+
+  for (const auto& guid : relay_header.to()) {
+    const auto p = vertical_handler_->find(guid_to_guid(guid));
+    if (p != vertical_handler_->end()) {
+      for (const auto& addr : p->second) {
+        vertical_handler_->enqueue_message(addr, msg);
```

Comment:
This pattern (for-each loop around `enqueue_message`) happens a few times.  Would it be cleaner to provide an `enqueue_message` overload that takes an AddressSet and does the loop internally?